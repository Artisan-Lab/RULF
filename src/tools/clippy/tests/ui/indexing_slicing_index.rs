#![feature(inline_const)]
#![warn(clippy::indexing_slicing)]
// We also check the out_of_bounds_indexing lint here, because it lints similar things and
// we want to avoid false positives.
#![warn(clippy::out_of_bounds_indexing)]
#![allow(unconditional_panic, clippy::no_effect, clippy::unnecessary_operation)]

const ARR: [i32; 2] = [1, 2];
const REF: &i32 = &ARR[idx()]; // Ok, should not produce stderr.
const REF_ERR: &i32 = &ARR[idx4()]; // Ok, let rustc handle const contexts.

const fn idx() -> usize {
    1
}
const fn idx4() -> usize {
    4
}

fn main() {
    let x = [1, 2, 3, 4];
    let index: usize = 1;
    x[index];
    x[4]; // Ok, let rustc's `unconditional_panic` lint handle `usize` indexing on arrays.
    x[1 << 3]; // Ok, let rustc's `unconditional_panic` lint handle `usize` indexing on arrays.

    x[0]; // Ok, should not produce stderr.
    x[3]; // Ok, should not produce stderr.
    x[const { idx() }]; // Ok, should not produce stderr.
    x[const { idx4() }]; // Ok, let rustc's `unconditional_panic` lint handle `usize` indexing on arrays.
    const { &ARR[idx()] }; // Ok, should not produce stderr.
    const { &ARR[idx4()] }; // Ok, let rustc handle const contexts.

    let y = &x;
    y[0]; // Ok, referencing shouldn't affect this lint. See the issue 6021
    y[4]; // Ok, rustc will handle references too.

    let v = vec![0; 5];
    v[0];
    v[10];
    v[1 << 3];

    const N: usize = 15; // Out of bounds
    const M: usize = 3; // In bounds
    x[N]; // Ok, let rustc's `unconditional_panic` lint handle `usize` indexing on arrays.
    x[M]; // Ok, should not produce stderr.
    v[N];
    v[M];
}
