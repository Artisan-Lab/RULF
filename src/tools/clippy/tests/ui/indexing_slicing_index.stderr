error[E0080]: evaluation of `main::{constant#3}` failed
  --> $DIR/indexing_slicing_index.rs:31:14
   |
LL |     const { &ARR[idx4()] }; // Ok, let rustc handle const contexts.
   |              ^^^^^^^^^^^ index out of bounds: the length is 2 but the index is 4

error[E0080]: erroneous constant used
  --> $DIR/indexing_slicing_index.rs:31:5
   |
LL |     const { &ARR[idx4()] }; // Ok, let rustc handle const contexts.
   |     ^^^^^^^^^^^^^^^^^^^^^^ referenced constant has errors

error: indexing may panic
  --> $DIR/indexing_slicing_index.rs:22:5
   |
LL |     x[index];
   |     ^^^^^^^^
   |
   = help: consider using `.get(n)` or `.get_mut(n)` instead
   = note: `-D clippy::indexing-slicing` implied by `-D warnings`

error: indexing may panic
  --> $DIR/indexing_slicing_index.rs:38:5
   |
LL |     v[0];
   |     ^^^^
   |
   = help: consider using `.get(n)` or `.get_mut(n)` instead

error: indexing may panic
  --> $DIR/indexing_slicing_index.rs:39:5
   |
LL |     v[10];
   |     ^^^^^
   |
   = help: consider using `.get(n)` or `.get_mut(n)` instead

error: indexing may panic
  --> $DIR/indexing_slicing_index.rs:40:5
   |
LL |     v[1 << 3];
   |     ^^^^^^^^^
   |
   = help: consider using `.get(n)` or `.get_mut(n)` instead

error: indexing may panic
  --> $DIR/indexing_slicing_index.rs:46:5
   |
LL |     v[N];
   |     ^^^^
   |
   = help: consider using `.get(n)` or `.get_mut(n)` instead

error: indexing may panic
  --> $DIR/indexing_slicing_index.rs:47:5
   |
LL |     v[M];
   |     ^^^^
   |
   = help: consider using `.get(n)` or `.get_mut(n)` instead

error[E0080]: evaluation of constant value failed
  --> $DIR/indexing_slicing_index.rs:10:24
   |
LL | const REF_ERR: &i32 = &ARR[idx4()]; // Ok, let rustc handle const contexts.
   |                        ^^^^^^^^^^^ index out of bounds: the length is 2 but the index is 4

error: aborting due to 9 previous errors

For more information about this error, try `rustc --explain E0080`.
